watchtower-plugin/src/wt_client.rs

```diff
@@ -0,0 +1,424 @@
+        &mut self,
+        tower_id: TowerId,
+        locator: Locator,
+        available_slots: u32,
+        receipt: &AppointmentReceipt,
+    ) {
+        if let Some(tower) = self.towers.get_mut(&tower_id) {
+            // DISCUSS: It may be nice to independently compute the slots and compare
+            tower.available_slots = available_slots;
+
+            self.dbm
+                .lock()
+                .unwrap()
+                .store_appointment_receipt(tower_id, locator, available_slots, receipt)
+                .unwrap();
+        } else {
+            log::error!(
+                "Cannot add appointment receipt to tower. Unknown tower_id: {}",
+                tower_id
+            );
+        }
+    }
+
+    /// Adds a pending appointment to the tower record.
+    pub fn add_pending_appointment(&mut self, tower_id: TowerId, appointment: &Appointment) {
+        if let Some(tower) = self.towers.get_mut(&tower_id) {
+            tower.pending_appointments.insert(appointment.locator);
+
+            self.dbm
+                .lock()
+                .unwrap()
+                .store_pending_appointment(tower_id, appointment)
+                .unwrap();
+        } else {
+            log::error!(
+                "Cannot add pending appointment to tower. Unknown tower_id: {}",
+                tower_id
+            );
+        }
+    }
+
+    /// Adds an invalid appointment to the tower record.
+    pub fn add_invalid_appointment(&mut self, tower_id: TowerId, appointment: &Appointment) {
+        if let Some(tower) = self.towers.get_mut(&tower_id) {
+            tower.invalid_appointments.insert(appointment.locator);
+
+            self.dbm
+                .lock()
+                .unwrap()
+                .store_invalid_appointment(tower_id, appointment)
+                .unwrap();
+        } else {
+            log::error!(
+                "Cannot add invalid appointment to tower. Unknown tower_id: {}",
+                tower_id
+            );
+        }
+    }
+
+    /// Flags a given tower as misbehaving, storing the misbehaving proof in the database.
+    pub fn flag_misbehaving_tower(&mut self, tower_id: TowerId, proof: MisbehaviorProof) {
+        if let Some(tower) = self.towers.get_mut(&tower_id) {
+            self.dbm
+                .lock()
+                .unwrap()
+                .store_misbehaving_proof(tower_id, &proof)
+                .unwrap();
+            tower.status = TowerStatus::Misbehaving;
+        } else {
+            log::error!("Cannot flag tower. Unknown tower_id: {}", tower_id);
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    use teos_common::test_utils::{
+        generate_random_appointment, get_random_appointment_receipt,
+        get_random_registration_receipt, get_random_user_id,
+    };
+
+    #[tokio::test]
+    async fn test_add_update_load_tower() {
+        let tmp_path = &format!(".watchtower_{}/", get_random_user_id());
```

Comment:
Stopping a running test will leave the working directory contaminated.
[`std::env::temp_dir`](https://doc.rust-lang.org/std/env/fn.temp_dir.html) could be used for that.

Reply:
I'll look into this, because I don't like the current create/remove boilerplate added to these tests because of the dir management.

Reply:
Have seen this one before, pretty convenient.